kie-server-parent/kie-server-services/kie-server-services-openshift/src/main/java/org/kie/server/services/openshift/impl/storage/cloud/KieServerStateCloudRepository.java

```diff
@@ -0,0 +1,231 @@
+        if (!retrieveKieServerId(kieServerState).equals(serverId)) {
+            throw new IllegalArgumentException("Invalid KieServerId: Id does not match with KieServerState.");
+        }
+        createOrUpdate(serverId, kieServerState, true);
+    }
+
+    @Override
+    @NotNull
+    public KieServerState load(@NotNull String serverId) {
+        KieServerState kieServerState = null;
+
+        try (KubernetesClient client = createKubernetesClient()) {
+            ConfigMap cm = client.configMaps().withName(serverId).get();
+
+            if (cm == null) {
+                throw new IllegalStateException(("KieServerId: [" + serverId + "], not found. Please create an associated ConfigMap with configuration first."));
+            }
+
+            Map<String, String> data = cm.getData();
+            kieServerState = (KieServerState) xs.fromXML(data.get(CFG_MAP_DATA_KEY));
+
+        } catch (IllegalStateException ise) {
+            logger.error("Loading KieServerState failed.", ise);
+            throw ise;
+        } catch (Exception e) {
+            logger.error("Loading KieServerState failed.", e);
+        }
+
+        if (kieServerState == null || !retrieveKieServerId(kieServerState).equals(serverId)) {
+            throw new IllegalStateException("Invalid KieServerId: " + serverId + ", or inconsistent state data.");
+        }
+
+        return kieServerState;
+    }
+
+    public List<String> retrieveAllKieServerIds() {
+        List<String> ids = null;
+        try (KubernetesClient client = createKubernetesClient()) {
+            ids = client.configMaps().list().getItems()
+                    .stream()
+                    .map(cfg -> cfg.getMetadata().getName())
+                    .collect(Collectors.toList());
+        } catch (Exception e) {
+            logger.error("Retrieve all KieServer ids failed.", e);
+        }
+        return ids;
+    }
+
+    public List<KieServerState> retrieveAllKieServerStates() {
+        List<KieServerState> states = null;
+        try (KubernetesClient client = createKubernetesClient()) {
+            states = client.configMaps().list().getItems()
+                    .stream()
+                    .map(cfg -> (KieServerState) xs.fromXML(cfg.getData().get(CFG_MAP_DATA_KEY)))
+                    .collect(Collectors.toList());
+        } catch (Exception e) {
+            logger.error("Retrieve KieServer id failed.", e);
+        }
+        return states;
+    }
+
+    public boolean exists(String id) {
+        boolean hasIt = false;
+        try (KubernetesClient client = createKubernetesClient()) {
+            hasIt = client.configMaps().withName(id).get() != null;
+        } catch (Exception e) {
+            logger.error("Search KieServerState with id [{}] failed.", id);
+        }
+        return hasIt;
+    }
+
+    public KieServerState delete(String id) {
+        KieServerState state = load(id);
+
+        try (OpenShiftClient client = createOpenShiftClient()) {
```

Comment:
Refactored code into two classes: KieServerStateCloudRepository which is compatible with vanilla Kubernetes cluster; KieServerStateOpenShiftRepository which includes this method with dependency on OpenShift client since DeploymentConfig is OpenShift only resource type.

Couple of reasons why defines KubernetesClient as a local variable:
1. Not quite confident on thread-safeness of fabric8 client
2. Call volume expected to these methods is low
3. To allow using 'try-with-resource' which is a cleaner and safer way of avoiding un-closed clients.
4. The performance concern of creating a OpenShift/K8S client should be really at Fabric8 side. It should employ common way of improve performance such as connection pooling etc.